File: WinForms/DevExpress.WindowsMailClient.Win/VB/DevExpress.WindowsMailClient.Win/Modules/Folders/MailViewControl/MailPresenter.vb
Imports System
Imports System.Drawing
Imports System.Windows.Forms
Imports DevExpress.Utils.Design
Imports DevExpress.Utils.Drawing
Imports DevExpress.WindowsMailClient.Win.Model
Imports DevExpress.XtraEditors
Imports DevExpress.XtraLayout.Utils

Namespace DevExpress.WindowsMailClient.Win.Modules

    Public Partial Class MailPresenter
        Inherits XtraUserControl

        Public Sub New()
            InitializeComponent()
            statusPanelItem.Visibility = LayoutVisibility.Never
            AddHandler LookAndFeel.ActiveLookAndFeel.StyleChanged, AddressOf OnLookAndFeelStyleChanged
            timer = CreateTimer()
        End Sub

        Protected Overrides Sub OnLoad(ByVal e As EventArgs)
            MyBase.OnLoad(e)
            If Not DesignTimeTools.IsDesignMode Then UpdateChildControlsAppearances()
        End Sub

        Protected Overrides Sub Dispose(ByVal disposing As Boolean)
            If disposing Then
                If Not DesignTimeTools.IsDesignMode Then RemoveHandler LookAndFeel.ActiveLookAndFeel.StyleChanged, AddressOf OnLookAndFeelStyleChanged
                If timer IsNot Nothing Then
                    timer.Stop()
                    timer.Dispose()
                    timer = Nothing
                End If
            End If

            MyBase.Dispose(disposing)
        End Sub

        Private Sub OnLookAndFeelStyleChanged(ByVal sender As Object, ByVal e As EventArgs)
            UpdateChildControlsAppearances()
        End Sub

        Protected Sub UpdateChildControlsAppearances()
            richEdit.Appearance.Text.ForeColor = ForeColor
            richEdit.Document.SetPageBackground(BackColor)
            mailStatusPanel1.SetColors(BackColor, ForeColor)
        End Sub

        Private message As Data.Message

        Private timer As Timer

        Private Function CreateTimer() As Timer
            Dim tmr As Timer = New Timer()
            tmr.Interval = 100
            AddHandler tmr.Tick, AddressOf OnTimerTick
            Return tmr
        End Function

        Private Sub OnTimerTick(ByVal sender As Object, ByVal e As EventArgs)
            timer.Stop()
            ShowMessageCore()
        End Sub

        Protected Friend Overridable Sub ShowMessage(ByVal message As Data.Message)
            If Me.message Is message Then Return
            Me.message = message
            If timer IsNot Nothing Then
                timer.Stop()
                timer.Start()
            End If
        End Sub

        Private Sub ShowMessageCore()
            If message Is Nothing Then
                ResetPresenterUI()
                Return
            End If

            transitionManager1.StartTransition(Me)
            Try
                labelSubject.Visibility = LayoutVisibility.Always
                labelFrom.Visibility = LayoutVisibility.Always
                labelDate.Visibility = LayoutVisibility.Always
                labelSubject.Text = message.Subject
                labelFrom.Text = message.From
                labelDate.Text = message.Date.ToString()
                Dim isHighPriority As Boolean = message.Priority = 2
                statusPanelItem.Visibility = If(isHighPriority, LayoutVisibility.Always, LayoutVisibility.Never)
                Dim photo As Image = MailClientDataModel.GetPhotoByEmail(message.Email)
                If photo Is Nothing Then
                    pePhoto.Image = MailClientDataModel.CreateSenderImage(message.From, pePhoto.Size, pePhoto.LookAndFeel.ActiveLookAndFeel)
                Else
                    pePhoto.Image = photo
                End If

                richEdit.MhtText = message.Text
                UpdateChildControlsAppearances()
            Finally
                transitionManager1.EndTransition()
            End Try
        End Sub

        Private Sub ResetPresenterUI()
            labelSubject.Text = Nothing
            labelFrom.Text = Nothing
            labelDate.Text = Nothing
            labelSubject.Visibility = LayoutVisibility.Never
            labelFrom.Visibility = LayoutVisibility.Never
            labelDate.Visibility = LayoutVisibility.Never
            statusPanelItem.Visibility = LayoutVisibility.Never
            pePhoto.Image = Nothing
            richEdit.MhtText = Nothing
            UpdateChildControlsAppearances()
        End Sub
    End Class

    Public Class MailStatusPanel
        Inherits Control

        Public Sub New()
            SetStyle(ControlStyles.AllPaintingInWmPaint Or ControlStyles.OptimizedDoubleBuffer, True)
            label = New LabelControl()
            label.Text = message
            label.BackColor = Color.Transparent
            label.Appearance.Font = New Font("Segoe UI", 8.25F)
            Controls.Add(label)
            OnLayoutChanged()
        End Sub

        Private label As LabelControl

        Private horzPadding As Integer = 20

        Private panelSize As Integer = 5

        Protected Overrides Sub ScaleControl(ByVal factor As SizeF, ByVal specified As BoundsSpecified)
            MyBase.ScaleControl(factor, specified)
            panelSize = CInt(factor.Width * panelSize)
            horzPadding = CInt(factor.Width * horzPadding)
            OnLayoutChanged()
        End Sub

        Const message As String = "This message was sent with high importance."

        Private _backColor As Color = Color.Empty

        Private _foreColor As Color = Color.Black

        Private lineColor As Color = Color.Empty

        Public Sub SetColors(ByVal backColor As Color, ByVal foreColor As Color)
            _backColor = MixColors(backColor, foreColor, 0.9)
            label.ForeColor = MixColors(backColor, foreColor, 0.25)
            lineColor = MixColors(backColor, foreColor, 0.65)
        End Sub

        Protected Overrides Sub OnSizeChanged(ByVal e As EventArgs)
            MyBase.OnSizeChanged(e)
            OnLayoutChanged()
            Invalidate()
        End Sub

        Private Sub OnLayoutChanged()
            label.Location = New Point(horzPadding, (Height - label.Height) \ 2)
        End Sub

        Protected Function MixColors(ByVal backColor As Color, ByVal foreColor As Color, ByVal value As Double) As Color
            Dim r As Integer = CInt(value * (backColor.R - foreColor.R) + foreColor.R)
            Dim g As Integer = CInt(value * (backColor.G - foreColor.G) + foreColor.G)
            Dim b As Integer = CInt(value * (backColor.B - foreColor.B) + foreColor.B)
            Return Color.FromArgb(r, g, b)
        End Function

        Protected Overrides Sub OnPaint(ByVal e As PaintEventArgs)
            MyBase.OnPaint(e)
            Using cache As GraphicsCache = New GraphicsCache(e)
                e.Graphics.Clear(_backColor)
                cache.FillRectangle(lineColor, New Rectangle(0, 0, panelSize, Height))
            End Using
        End Sub
    End Class
End Namespace

File: WinForms/DevExpress.WindowsMailClient.Win/CS/DevExpress.WindowsMailClient.Win/Modules/Folders/MailViewControl/MailPresenter.cs
using System;
using System.Drawing;
using System.Windows.Forms;
using DevExpress.Utils.Design;
using DevExpress.Utils.Drawing;
using DevExpress.WindowsMailClient.Win.Model;
using DevExpress.XtraEditors;
using DevExpress.XtraLayout.Utils;

namespace DevExpress.WindowsMailClient.Win.Modules {
    public partial class MailPresenter : XtraUserControl {
        public MailPresenter() {
            InitializeComponent();
            this.statusPanelItem.Visibility = LayoutVisibility.Never;
            LookAndFeel.ActiveLookAndFeel.StyleChanged += OnLookAndFeelStyleChanged;
            this.timer = CreateTimer();
        }
        protected override void OnLoad(EventArgs e) {
            base.OnLoad(e);
            if(!DesignTimeTools.IsDesignMode)
                UpdateChildControlsAppearances();
        }
        protected override void Dispose(bool disposing) {
            if(disposing) {
                if(!DesignTimeTools.IsDesignMode)
                    LookAndFeel.ActiveLookAndFeel.StyleChanged -= OnLookAndFeelStyleChanged;
                if(this.timer != null) {
                    this.timer.Stop();
                    this.timer.Dispose();
                    this.timer = null;
                }
            }
            base.Dispose(disposing);
        }
        void OnLookAndFeelStyleChanged(object sender, EventArgs e) {
            UpdateChildControlsAppearances();
        }
        protected void UpdateChildControlsAppearances() {
            richEdit.Appearance.Text.ForeColor = ForeColor;
            richEdit.Document.SetPageBackground(BackColor);
            mailStatusPanel1.SetColors(BackColor, ForeColor);
        }
        Data.Message message;
        Timer timer;
        Timer CreateTimer() {
            Timer tmr = new Timer();
            tmr.Interval = 100;
            tmr.Tick += OnTimerTick;
            return tmr;
        }

        void OnTimerTick(object sender, EventArgs e) {
            this.timer.Stop();
            ShowMessageCore();
        }
        protected internal virtual void ShowMessage(Data.Message message) {
            if(this.message == message)
                return;
            this.message = message;
            if(this.timer != null) {
                this.timer.Stop();
                this.timer.Start();
            }
        }
        void ShowMessageCore() {
            if(this.message == null) {
                ResetPresenterUI();
                return;
            }
            this.transitionManager1.StartTransition(this);
            try {
                labelSubject.Visibility = LayoutVisibility.Always;
                labelFrom.Visibility = LayoutVisibility.Always;
                labelDate.Visibility = LayoutVisibility.Always;

                this.labelSubject.Text = message.Subject;
                this.labelFrom.Text = message.From;
                this.labelDate.Text = message.Date.ToString();
                bool isHighPriority = message.Priority == 2;
                this.statusPanelItem.Visibility = isHighPriority ? LayoutVisibility.Always : LayoutVisibility.Never;
                Image photo = MailClientDataModel.GetPhotoByEmail(message.Email);
                

                if(photo == null)
                    pePhoto.Image = MailClientDataModel.CreateSenderImage(message.From, pePhoto.Size, pePhoto.LookAndFeel.ActiveLookAndFeel);
                else this.pePhoto.Image = photo;
                this.richEdit.MhtText = message.Text;
                UpdateChildControlsAppearances();
            }
            finally {
                this.transitionManager1.EndTransition();
            }
        }
        void ResetPresenterUI() {
            this.labelSubject.Text = null;
            this.labelFrom.Text = null;
            this.labelDate.Text = null;
            labelSubject.Visibility = LayoutVisibility.Never;
            labelFrom.Visibility = LayoutVisibility.Never;
            labelDate.Visibility = LayoutVisibility.Never;
            this.statusPanelItem.Visibility = LayoutVisibility.Never;
            pePhoto.Image = null;
            this.richEdit.MhtText = null;
            UpdateChildControlsAppearances();
        }
    }
    public class MailStatusPanel : Control {
        public MailStatusPanel() {
            SetStyle(ControlStyles.AllPaintingInWmPaint | ControlStyles.OptimizedDoubleBuffer, true);
            this.label = new LabelControl();
            this.label.Text = message;
            this.label.BackColor = Color.Transparent;
            this.label.Appearance.Font = new Font("Segoe UI", 8.25f);
            Controls.Add(label);
            OnLayoutChanged();
        }
        LabelControl label;
        int horzPadding = 20;
        int panelSize = 5;
        protected override void ScaleControl(SizeF factor, BoundsSpecified specified) {
            base.ScaleControl(factor, specified);
            panelSize = (int)(factor.Width * panelSize);
            horzPadding = (int)(factor.Width * horzPadding);
            OnLayoutChanged();
        }
        const string message = "This message was sent with high importance.";
        Color _backColor = Color.Empty;
        Color _foreColor = Color.Black;
        Color lineColor = Color.Empty;
        public void SetColors(Color backColor, Color foreColor) {
            this._backColor = MixColors(backColor, foreColor, 0.9);
            this.label.ForeColor = MixColors(backColor, foreColor, 0.25);
            this.lineColor = MixColors(backColor, foreColor, 0.65);
        }
        protected override void OnSizeChanged(EventArgs e) {
            base.OnSizeChanged(e);
            OnLayoutChanged();
            Invalidate();
        }

        void OnLayoutChanged() {
            this.label.Location = new Point(horzPadding, (Height - label.Height) / 2);
        }
        protected Color MixColors(Color backColor, Color foreColor, double value) {
            int r = (int)(value * (backColor.R - foreColor.R) + foreColor.R);
            int g = (int)(value * (backColor.G - foreColor.G) + foreColor.G);
            int b = (int)(value * (backColor.B - foreColor.B) + foreColor.B);
            return Color.FromArgb(r, g, b);
        }
        protected override void OnPaint(PaintEventArgs e) {
            base.OnPaint(e);
            using(GraphicsCache cache = new GraphicsCache(e)) {
                e.Graphics.Clear(_backColor);
                cache.FillRectangle(lineColor, new Rectangle(0, 0, panelSize, Height));
            }
        }
    }
}
